### PR TITLE
Option to hide prompt in iOS + disposed semaphore.

### PR DESCRIPTION
**Pull request by SameerK-MSFT**
Added option to hide prompt in iOS. Disposed SemaphoreSlim after use to release resource.
#2131

### CODE REVIEW

**Review by trwalke — DISMISSED** [review]
Minor comment. 
LGTM otherwise.

**Comment by henrik-me** [comment]
no tests/utests? ... would be great to have system browser enabled for e2e testing....
imo: at minimum an existing manual test should be expanded to cover this scenario

File: build/platform_and_feature_flags.props [hunk 1]
```diff
@@ -6,19 +6,19 @@
     <DefineConstants>$(DefineConstants);NET5_WIN;SUPPORTS_CONFIDENTIAL_CLIENT;SUPPORTS_CUSTOM_CACHE;SUPPORTS_OS_SYSTEM_BROWSER;SUPPORTS_BROKER;SUPPORTS_WEBVIEW2</DefineConstants>
   </PropertyGroup>
   <PropertyGroup Condition="'$(TargetFramework)' == '$(TargetFrameworkUap)' ">
-    <DefineConstants>$(DefineConstants);WINDOWS_APP;SUPPORTS_BROKER;SUPPORTS_CUSTOM_CACHE</DefineConstants>
+    <DefineConstants>$(DefineConstants);WINDOWS_APP;SUPPORTS_BROKER;SUPPORTS_OS_SYSTEM_BROWSER;SUPPORTS_CUSTOM_CACHE</DefineConstants>
```
Comment (bgavrilMS) [inline build/platform_and_feature_flags.props:9]: UAP / UWP (Universal Windows Platform) does not support SYSTEM_BROWSER

File: src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs [hunk 2]
```diff
@@ -1,169 +1,176 @@
...
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public SystemWebViewOptions()
+        {
+            ValidatePlatformAvailability();
+        }
+
+        /// <summary>
+        /// When the user finishes authenticating, MSAL will respond with a 200 OK message,
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs:37]: Please add comments that these existing options are exclusive for desktop operating systems (so .NET Clasic and .NET Core). And if the user tries to `set` one of these on Android, iOS or UWP, we should throw a nice exception.

File: src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs [hunk 3]
```diff
@@ -1,169 +1,176 @@
...
+        /// When the user finishes authenticating, but an error occurred, MSAL will redirect the browser to the given Uri
+        /// </summary>
+        /// <remarks>Takes precedence over <see cref="HtmlMessageError"/></remarks>
+        public Uri BrowserRedirectError { get; set; }
+
+        /// <summary>
+        /// This hides privacy prompt displayed on the iOS Devices (ver 13.0+) when set to true.
+        /// By default, it is false and displays the prompt.
+        /// </summary>
+        public bool HidePrivacyPrompt { get; set; } = false;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs:65]: This one should only be visible in iOS and NetStandard. Trying to set this from another platform should result in an exception and the API should be hidden as well.

Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs:65]: For naming purposes ... I would name it `iOSHidePrivacyPrompt` to make it clear it only affects iOS. And the comments should mention prefersEphemeralWebBrowserSession and the fact that the browser will not remember previous logged-in users.

File: src/client/Microsoft.Identity.Client/Platforms/iOS/IosWebUIFactory.cs [hunk 4]
```diff
@@ -1,40 +1,41 @@
...
-            
-            return new SystemWebUI()
-            {
-                RequestContext = requestContext
-            };
-        }
-
-    }
-}
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Platforms/iOS/IosWebUIFactory.cs:1]: There's smth wrong with your spacing settings. The diffs are all screwed up :)

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Platforms/iOS/IosWebUIFactory.cs:1]: We use the VS defaults around tabs and there is an .editorconfig file that sets some styles for you... 
image

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
ios logic looks good, but needs some work to better hide some of the APIs

File: src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs [hunk 5]
```diff
@@ -1,169 +1,176 @@
...
+#if DESKTOP
+            url = url.Replace("&", "^&");
+            Process.Start(new ProcessStartInfo("cmd", $"/c start microsoft-edge:{url}") { CreateNoWindow = true });
+            await Task.FromResult(0).ConfigureAwait(false);
+#else
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                url = url.Replace("&", "^&");
+                Process.Start(new ProcessStartInfo("cmd", $"/c start microsoft-edge:{url}") { CreateNoWindow = true });
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs:122]: Becuase you included this in the Android and iOS, it causes a crash. Process is not available - see the PR build failure ... `src\client\Microsoft.Identity.Client\ApiConfig\SystemWebViewOptions.cs(122,17): Error CS0103: The name 'Process' does not exist in the current context`

File: build/platform_and_feature_flags.props (hunk 1 above)
Comment (SameerK-MSFT) [reply build/platform_and_feature_flags.props]: Updated to removed SUPPORTS_OS_SYSTEM_BROWSER from all configurations

File: src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs (hunk 3 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs]: Hides for UWP i.e. WINDOWS_APP

File: src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs (hunk 5 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs:122]: Updated the flags

File: src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs (hunk 2 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs:37]: Added the hide attribute

File: src/client/Microsoft.Identity.Client/Platforms/iOS/IosWebUIFactory.cs (hunk 4 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/Platforms/iOS/IosWebUIFactory.cs:1]: I have those settings. Not sure what went wrong. Hope it goes fine in the second push.

File: src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs (hunk 3 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/ApiConfig/SystemWebViewOptions.cs]: Good suggestion. Done.

**Review by bgavrilMS — APPROVED** [review]

**Comment by SameerK-MSFT** [comment]
> no tests/utests? ... would be great to have system browser enabled for e2e testing....
> imo: at minimum an existing manual test should be expanded to cover this scenario

I have a test ready in the sample app. That can be pushed after the new NuGet gets published and consumed by App.

**Comment by jennyf19** [comment]
@SameerK-MSFT the current iOS tests are using the [embedded webview](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/tests/devapps/XForms/XForms/TestPage.xaml.cs#L196), i don't remember exactly, but we don't have system browser for the UI automated tests. You should be able to make changes to the other Xamarin dev app now, which would apply the changes you made. Also, did you change the spaces to tabs because the diff still looks odd to me.

**Comment by bgavrilMS** [comment]
The iOS builds seem to be failing with an env error: 

`Xamarin.iOS task failed with error Failed to find iOS signing identity. Verify the signing and provisioning information provided.. For guidance on setting up the build definition, see https://go.microsoft.com/fwlink/?LinkId=760847.`

Is it possible that the test provisioning profile expired? @trwalke - is this the same error happening a few weeks ago?

**Comment by trwalke** [comment]
> The iOS builds seem to be failing with an env error:
> 
> `Xamarin.iOS task failed with error Failed to find iOS signing identity. Verify the signing and provisioning information provided.. For guidance on setting up the build definition, see https://go.microsoft.com/fwlink/?LinkId=760847.`
> 
> Is it possible that the test provisioning profile expired? @trwalke - is this the same error happening a few weeks ago?

Yes, However, this time is seems to be related to an actual issue with the cert and not just a service issue, im requesting a new certificate for the build.
@bgavrilMS

**Comment by trwalke** [comment]
restarting builds

**Comment by SameerK-MSFT** [comment]
I will update sample app after the package is released

**Pull request merged**